tasks/Task.cpp

```diff
@@ -42,8 +75,176 @@ void Task::errorHook()
 void Task::stopHook()
 {
     TaskBase::stopHook();
+    m_handle.reset();
 }
 void Task::cleanupHook()
 {
     TaskBase::cleanupHook();
 }
+
+sensor::sensor_info Task::getMetadata()
+{
+    auto metadata = sensor::get_metadata(*m_handle);
+    return sensor::parse_metadata(metadata);
+}
+
+static int recv_fixed(int fd, void* buf, int64_t packet_len, int64_t len)
+{
+    int64_t bytes_read = recv(fd, (char*)buf, len, 0);
+
+    if (bytes_read == 0 || bytes_read % packet_len != 0) {
+        return -1;
+    }
+    else {
+        return bytes_read;
+    }
+}
+
+LidarScan Task::acquireData()
+{
+    size_t width = m_metadata.format.columns_per_frame;
+    size_t heigth = m_metadata.format.pixels_per_column;
+
+    // A LidarScan holds lidar data for an entire rotation of the device
+    LidarScan scan{width, heigth, m_metadata.format.udp_profile_lidar};
+
+    if (m_buffer_bytes) {
+        if (processBuffer(scan, m_buffer_offset, m_buffer_bytes)) {
+            return scan;
+        }
+    }
+
+    base::Time deadline = base::Time::now() + base::Time::fromSeconds(m_batcher_timeout);
+
+    while (base::Time::now() < deadline) {
+        // TODO: update poll_client timeout to reflect the remaining time
+        sensor::client_state cli_state = sensor::poll_client(*m_handle);
+        // check for error status
+        if (cli_state & sensor::CLIENT_ERROR) {
+            throw std::runtime_error("Sensor client returned error state!");
+        }
+        // check for lidar data, read a packet and add it to the current batch
+        if (cli_state & sensor::LIDAR_DATA) {
+            int bytes_read = recv_fixed(m_handle->lidar_fd,
+                m_packet_buffer.data(),
+                m_packet_format->lidar_packet_size,
+                m_packet_buffer.size());
+            if (bytes_read < 0) {
+                throw std::runtime_error("Failed to read a packet of the expected size!");
+            }
+            if (processBuffer(scan, 0, bytes_read)) {
+                return scan;
+            }
+        }
+        this_thread::sleep_for(microseconds(2000));
```

Comment:
I dont understand the reasoning behind this sleep? Is it to allow the client to process more data? Maybe add a comment explaining it

Reply:
not used anymore